search/materialize.go

```diff
@@ -275,35 +270,79 @@ func totalRows(rr []RowRange) int64 {
 	return res
 }
 
-func (m *Materializer) materializeChunks(ctx context.Context, rgi int, mint, maxt int64, rr []RowRange) ([][]chunks.Meta, error) {
+func (m *Materializer) materializeChunks(rgi int, mint, maxt int64, rr []RowRange) chunkIterableSet {
 	minDataCol := m.s.DataColumIdx(mint)
 	maxDataCol := m.s.DataColumIdx(maxt)
 	rg := m.b.ChunksFile().RowGroups()[rgi]
-	r := make([][]chunks.Meta, totalRows(rr))
 
-	for i := minDataCol; i <= min(maxDataCol, len(m.dataColToIndex)-1); i++ {
-		values, err := m.materializeColumn(ctx, m.b.ChunksFile(), rg, rg.ColumnChunks()[m.dataColToIndex[i]], rr)
-		if err != nil {
-			return r, err
-		}
+	dataColCount := min(maxDataCol, len(m.dataColToIndex)-1) - minDataCol + 1
 
-		for vi, value := range values {
-			chks, err := m.d.Decode(value.ByteArray(), mint, maxt)
-			if err != nil {
-				return r, errors.Wrap(err, "failed to decode chunks")
-			}
-			r[vi] = append(r[vi], chks...)
-		}
+	if dataColCount <= 0 {
+		return &repeatEmptyChunkIterableSet{count: totalRows(rr)}
 	}
 
-	return r, nil
+	result := &chunksIteratorIterator{
+		d:    m.d,
+		mint: mint,
+		maxt: maxt,
+	}
+
+	for colIdx := range dataColCount {
+		iter := m.materializeColumn(m.ctx, m.b.ChunksFile(), rg, rg.ColumnChunks()[m.dataColToIndex[minDataCol+colIdx]], rr)
+		result.colIts = append(result.colIts, iter)
+	}
+	return result
 }
 
-func (m *Materializer) materializeColumn(ctx context.Context, file *storage.ParquetFile, group parquet.RowGroup, cc parquet.ColumnChunk, rr []RowRange) ([]parquet.Value, error) {
+func (m *Materializer) materializeColumn(ctx context.Context, file *storage.ParquetFile, group parquet.RowGroup, cc parquet.ColumnChunk, rr []RowRange) *columnIterator {
 	if len(rr) == 0 {
-		return nil, nil
+		return &columnIterator{}
+	}
+
+	pageBatches, err := m.getPageRanges(group, cc, rr)
+	if err != nil {
+		return &columnIterator{
+			err: errors.Wrap(err, "failed to get page ranges"),
+		}
+	}
+
+	return &columnIterator{
+		ctx:         ctx,
+		file:        file,
+		cc:          cc,
+		pageBatches: pageBatches,
+		iter: &pagesBatchIterator{
+			pi: new(pageIterator),
+		},
+	}
+}
+
+// materializeColumnAsSlice internally calls materializeColumn and collects all values into a slice.
+// It is convenient for callers who don't want to stream values.
+func (m *Materializer) materializeColumnAsSlice(ctx context.Context, file *storage.ParquetFile, group parquet.RowGroup, cc parquet.ColumnChunk, rr []RowRange) ([]parquet.Value, error) {
+	iter := m.materializeColumn(ctx, file, group, cc, rr)
+	values := make([]parquet.Value, 0, totalRows(rr))
```

Comment:
Maybe this slice could also be subject for memory pooling in the future. Not necessary now.

Reply:
Agreed. There's a couple of other places where we could do that too. But yeah I deliberately didn't to avoid making this change bigger and (even more) complex. We can have a followup PR that adds pooling everywhere we see it possible, maybe in the one you've already started https://github.com/prometheus-community/parquet-common/pull/69